### PR TITLE
Refactors for `Boxed` and `Option`

### LINKED ISSUE #828: [Bug] Wrapping an enum parameter in Option
### Describe the bug

Looks like I might've ran into a bug with code gen:

```rust
pub enum BugTest {
    Foobar,
}
pub fn bug_test(foo: Option<BugTest>) {}
```

 Yields the following output (I have this with a `build.rs` if it matters):

Screenshot

This happens only when wrapping the enum in `Option`. Passing the enum directly without `Option` runs ok.

Is this intentional behavior or is it an issue with codegen? I would love to contribute with a PR, but I unfortunately only started learning rust a few days ago, and would have no clue where to start.

### Codegen logs with `RUST_LOG=debug` environment variable

```shell
Output too long to paste here, so here's a link:

https://pastebin.com/uDJ71iS7
```

### To Reproduce

```rust
pub enum BugTest {
    Foobar,
}
pub fn bug_test(foo: Option<BugTest>) {}
```

### Expected behavior

_No response_

### Generated binding code

_No response_

### OS

macOS

### Version of `flutter_rust_bridge_codegen`

1.49.1

### Flutter info

_No response_

### Version of `clang++`

_No response_

### Version of `ffigen`

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Desdaemon**
- Refactors `EnumRef` to make `Option<FieldlessEnum>` work
- (Potentially breaking) `Box<primitive>` and `Option<Box<primitive>>` on WASM no longer allocates a Box, but receives a (nullable) value directly from Dart. Makes the implementation cleaner.

Closes #828.

## Checklist

- [x] An issue to be fixed by this PR is listed above.
- [x] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [x] The code generator is run and the code is formatted (e.g. via `just refresh_all`).
- [x] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [x] CI is passing.

### CODE REVIEW

**Comment by fzyzcjy** [comment]
Good job! Will review after CI passes

**Comment by Desdaemon** [comment]
> Good job! Will review after CI passes

The last CI failure should be the README audit, as one of the links began with `http://`. What would be the best way to fix that?

**Comment by fzyzcjy** [comment]
Just change it to https manually

**Comment by fzyzcjy** [comment]
Good, I will review in a few hours :)

File: frb_codegen/src/generator/dart/ty_boxed.rs
Comment (fzyzcjy) [inline frb_codegen/src/generator/dart/ty_boxed.rs:65]: nit (unrelated to this line of code): so, have our tests covered *all* cases that this PR changes?

File: frb_codegen/src/ir/ty.rs
Comment (fzyzcjy) [inline frb_codegen/src/ir/ty.rs:130]: uber-nit: seems that your formatter does not agree with previous contributor's ;) so, shall we somehow create a stricter linter rule that everyone has to follow, thus avoiding such code "change" - which will make `git blame` and code review harder

**Review by fzyzcjy — COMMENTED** [review]
Good job! I am very busy with my research currently, so have not had a deep review about this. However, given your reputation - your wonderful contributions to these components and the library - I believe in you :) I may do a brief review in a day when I have some energy

File: frb_codegen/src/generator/rust/ty_optional.rs
Comment (Desdaemon) [inline frb_codegen/src/generator/rust/ty_optional.rs:42]: It's been a while, but this hack made me think if it was wiser to have Optional act more like a delegate, so that a nested IrType tree wouldn't be necessary.

File: frb_codegen/src/ir/ty.rs
Comment (Desdaemon) [reply frb_codegen/src/ir/ty.rs:130]: My bad, I'll try to revert some of these nonessential changes.

**Comment by Desdaemon** [comment]
Also good luck with your research @fzyzcjy!

File: frb_example/pure_dart/rust/src/api.rs
Comment (fzyzcjy) [reply frb_example/pure_dart/rust/src/api.rs:921]: Quite weird...

**Comment by fzyzcjy** [comment]
Btw is #901 somehow related?

**Comment by fzyzcjy** [comment]
> Also good luck with your research @fzyzcjy!

Thank you!

File: frb_codegen/src/ir/ty.rs
Comment (fzyzcjy) [reply frb_codegen/src/ir/ty.rs]: It's OK. The main problem is, we may need to somehow setup a consistent formatting rule. You know, for example, in Dart there is one consistent formatter so everyone's code is the same

**Review by fzyzcjy — APPROVED** [review]
Briefly reviewed, good job! Now just wait for CI to pass

**Comment by Desdaemon** [comment]
> Btw is #901 somehow related?

At first glance, not much. That PR's changes seem to be around structs instead.

**Comment by fzyzcjy** [comment]
Hmm ios CI seems to be reliably failing now (I have triggered a retry). Will merge after it passes

**Comment by Desdaemon** [comment]
@fzyzcjy Should be ready to merge now.

**Comment by fzyzcjy** [comment]
Great job!

**Pull request merged**
